src/icmp/Icmp.h

```diff
@@ -16,8 +16,8 @@
 #define PINGER_PAYLOAD_SZ   8192
 
 #define MAX_PAYLOAD 256 // WAS: SQUIDHOSTNAMELEN
-#define MAX_PKT4_SZ (MAX_PAYLOAD + sizeof(struct timeval) + sizeof (char) + sizeof(struct icmphdr) + 1)
-#define MAX_PKT6_SZ (MAX_PAYLOAD + sizeof(struct timeval) + sizeof (char) + sizeof(struct icmp6_hdr) + 1)
+#define MAX_PKT4_SZ (MAX_PAYLOAD + sizeof(struct timeval) + sizeof (char) + sizeof(struct icmphdr) + sizeof(struct icmpEchoData) + 1)
+#define MAX_PKT6_SZ (MAX_PAYLOAD + sizeof(struct timeval) + sizeof (char) + sizeof(struct icmp6_hdr) + sizeof(struct icmpEchoData) + 1)
```

Comment:
The on-wire size of icmpEchoData is already accounted for in these calculations.
```
struct icmpEchoData {
    struct timeval tv;    // sizeof(struct timeval)
    unsigned char opcode; // sizeof(char)
    char payload[MAX_PAYLOAD];
};
```
What could be going on is compiler padding and/or shuffling the structure members around to "optimal" byte positions that do not match the on-wire bits.

A static_assert should be added after struct icmpEchoData definition to verify that struct actually matches the on-wire bitsize. Then we can clarify the MAX_PKT*_SZ definitions:

```
#define MAX_PKT4_SZ (sizeof(struct icmphdr) + sizeof(struct icmpEchoData) + 1)
#define MAX_PKT6_SZ (sizeof(struct icmp6_hdr) + sizeof(struct icmpEchoData) + 1)
```

Reply:
@yadij, do you know what that `+1` is for? Perhaps we should drop it while we are fixing these constants?

Reply:
Thanks for the reply and the clarification, @yadij.  Indeed, this is exactly what seems to be happening.  I've added a `__attribute__((packed))` at the definition of `icmpEchoData`, and the build passed.

If I discard the modification to `MAX_PKT{4,6}_SZ` and just add a `static_assert` after the structure definition, then obviously the assert triggers (with the on-wire bitsize being `sizeof(struct timeval) + sizeof(char) + MAX_PAYLOAD`).  It seems to me that using `__attribute__((packed))` would be the less intrusive change in this case.

Reply:
Of course, if having padding on `icmpEchoData` is not a problem (as it seems to be the case based on what @rousskov said below), then another option would be to relax the `static_assert` (like `sizeof(icmpEchoData) >= sizeof all fields)`) and just leave it like that.  I understand that using `__attribute__((packed))` may be frowned upon because it's GCC-specific.

Reply:
> using `__attribute__((packed))` may be frowned upon because it's GCC-specific.

Yes, we should not use non-standard and/or compiler-specific language extensions when general solutions are within reach.

> `sizeof(icmpEchoData) >= sizeof all fields`

We do not need a static assertion that protects us from insanely broken compilers IMO.

We should have a static assertion that protects us from somebody accidentally making the fixed-by-this-PR `MAX_PKT{4,6}_SZ` value smaller again. To avoid looking silly, that assertion should probably be placed in Icmp4::SendEcho() code and use sizeof(pkt) rather than `MAX_PKT4_SZ`. Something along these lines may work:

```C++
static_assert(sizeof(*icmp) + sizeof(*echo) <= sizeof(pkt), "our custom ICMP Echo payload fits the packet buffer");
```

Technically, we must also rework the `(icmp + 1)` code for the same reasons because it is susceptible to post-cmphdr padding in an array context. Our sizeof+sizeof math and assertion assume no such padding... A safer version would cast `icmp` pointer to `char*` and then add `sizeof(*icmp)` instead, before casting to icmpEchoData. Please use C++ casts.

Reply:
> @yadij, do you know what that `+1` is for? Perhaps we should drop it while we are fixing these constants?

I do not recall exactly. Probably a hack to ensure '\0' existence after the echo data, so c-string code does not over-read/write memory when data member is not a terminated.

Reply:
> Probably a hack to ensure '\0' existence after the echo data, so c-string code does not over-read/write memory when data member is not a terminated.

If that is true, there is probably a bug somewhere. No code should interpret raw echo data as a c-string AFAICT. It would be nice to investigate and remove `+1` (fixing any related code if needed), but I do not insist on that. Adding a corresponding TODO would be nice so that we do not need to ask this question again.